examples/drivers/windows/otLwf/tunnel.c

```diff
@@ -0,0 +1,1745 @@
+    {
+        LogError(DRIVER_DEFAULT, "Failed to query SPINEL_PROP_INTERFACE_TYPE, %!STATUS!", Status);
+        goto error;
+    }
+    if (InterfaceType != SPINEL_PROTOCOL_TYPE_THREAD)
+    {
+        Status = STATUS_NOT_SUPPORTED;
+        LogError(DRIVER_DEFAULT, "SPINEL_PROP_INTERFACE_TYPE is invalid, %d", InterfaceType);
+        goto error;
+    }
+
+    // TODO - Query other values and capabilities
+
+error:
+
+    if (!NT_SUCCESS(Status))
+    {
+        otLwfUninitializeTunnelMode(pFilter);
+    }
+
+    LogFuncExitNDIS(DRIVER_DEFAULT, Status);
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void 
+otLwfUninitializeTunnelMode(
+    _In_ PMS_FILTER pFilter
+    )
+{
+    LogFuncEntry(DRIVER_DEFAULT);
+
+    // Clean up worker thread
+    if (pFilter->TunWorkerThread)
+    {
+        LogInfo(DRIVER_DEFAULT, "Stopping tunnel worker thread and waiting for it to complete.");
+
+        // Send event to shutdown worker thread
+        KeSetEvent(&pFilter->TunWorkerThreadStopEvent, 0, FALSE);
+
+        // Wait for worker thread to finish
+        KeWaitForSingleObject(
+            pFilter->TunWorkerThread,
+            Executive,
+            KernelMode,
+            FALSE,
+            NULL
+            );
+
+        // Free worker thread
+        ObDereferenceObject(pFilter->TunWorkerThread);
+        pFilter->TunWorkerThread = NULL;
+
+        LogInfo(DRIVER_DEFAULT, "Tunnel worker thread cleaned up.");
+    }
+
+    // TODO - Clean up command handlers
+
+    LogFuncExit(DRIVER_DEFAULT);
+}
+
+// Worker thread for processing all tunnel events
+_Use_decl_annotations_
+VOID
+otLwfTunWorkerThread(
+    PVOID   Context
+    )
+{
+    PMS_FILTER pFilter = (PMS_FILTER)Context;
+    NT_ASSERT(pFilter);
+
+    LogFuncEntry(DRIVER_DEFAULT);
+
+    PKEVENT WaitEvents[] = 
+    { 
+        &pFilter->TunWorkerThreadStopEvent,
+        &pFilter->TunWorkerThreadAddressChangedEvent
+    };
+
+    LogFuncExit(DRIVER_DEFAULT);
+    
+    while (true)
+    {
+        // Wait for event to stop or process event to fire
+        NTSTATUS status = 
+            KeWaitForMultipleObjects(
+                ARRAYSIZE(WaitEvents), 
+                (PVOID*)WaitEvents, 
+                WaitAny, 
+                Executive, 
+                KernelMode, 
+                FALSE, 
+                NULL, 
+                NULL);
+
+        // If it is the first event, then we are shutting down. Exit loop and terminate thread
+        if (status == STATUS_WAIT_0)
+        {
+            LogInfo(DRIVER_DEFAULT, "Received tunnel worker thread shutdown event.");
+            break;
+        }
+        else if (status == STATUS_WAIT_0 + 1) // TunWorkerThreadAddressChangedEvent fired
+        {
+            PVOID DataBuffer = NULL;
+            const uint8_t* value_data_ptr = NULL;
+            spinel_size_t value_data_len = 0;
+            
+            // Query the current addresses
+            status = 
+                otLwfGetTunProp(
```

Comment:
It isn't entirely clear to me why `TunWorkerThread` would need a special case for `PROP_IPV6_ADDRESS_TABLE`.